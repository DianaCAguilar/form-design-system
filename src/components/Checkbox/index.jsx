import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { CustomCheckboxContainer, CustomCheckboxInput } from '@reach/checkbox';
import cc from 'classcat';

import CheckEmptyIcon from 'lib/icons/react/CheckEmptyIcon';
import CheckFilledIcon from 'lib/icons/react/CheckFilledIcon';
import CheckIndeterminateIcon from 'lib/icons/react/CheckIndeterminateIcon';

const Checkbox = ({ checked, label, disabled, showLabel = true, ...rest }) => {
  const CheckboxIcon = useMemo(() => {
    if (checked) {
      if (checked === 'mixed') {
        return CheckIndeterminateIcon;
      }
      return CheckFilledIcon; // true
    }
    return CheckEmptyIcon; // false
  }, [checked]);

  const Element = showLabel ? 'label' : 'div';

  return (
    <Element className="alignChild--left--center">
      <CustomCheckboxContainer
        checked={checked}
        disabled={disabled}
        className={cc([
          { 'fdsCheckbox--checked': checked, 'fdsCheckbox--disabled': disabled },
          'fdsCheckbox-container',
        ])}
        {...rest}
      >
        <CustomCheckboxInput
          disabled={disabled}
          checked={checked}
          aria-label={!showLabel ? label : undefined}
          {...rest}
        />
        <span aria-hidden className="fdsCheckbox-span">
          <CheckboxIcon size="xs" />
        </span>
      </CustomCheckboxContainer>
      {showLabel && <span className="fdsCheckbox-label">{label}</span>}
    </Element>
  );
};

Checkbox.propTypes = {
  /**
   * Setting any one of the 3 values here will turn component into controlled component.
   * */
  checked: PropTypes.bool,

  /** Label used for `label` element */
  label: PropTypes.string.isRequired,

  /** Disables form field when `true` */
  disabled: PropTypes.bool,

  /** Controls whether the label surfaces as a text label, or is hidden and applied as an aria-label */
  showLabel: PropTypes.bool,
};

export default Checkbox;
