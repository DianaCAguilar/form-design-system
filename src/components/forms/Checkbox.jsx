import React, { useState } from 'react';
import PropTypes from 'prop-types';
import FDS from 'lib/dictionary/js/styleConstants';

import CheckEmptyIcon from 'lib/icons/react/CheckEmptyIcon';
import CheckFilledIcon from 'lib/icons/react/CheckFilledIcon';
import CheckIndeterminateIcon from 'lib/icons/react/CheckIndeterminateIcon';

import { CustomCheckboxInput, CustomCheckboxContainer } from '@reach/checkbox';

const Checkbox = ({ label, defaultChecked, showLabel = true, ...rest }) => {
  const [checkedState, setChecked] = useState(rest.checked || defaultChecked || false);
  const checked = rest.checked != null ? rest.checked : checkedState;

  // We should maybe fire onChange in "addition" to setChecked, based on whether we
  // interpret checkbox as being "controlled" or not
  const onChange = rest.onChange ? rest.onChange : (e) => setChecked(e.target.checked);

  const CheckboxIcon = (() => {
    if (checked) {
      if (checked === 'mixed') {
        return CheckIndeterminateIcon;
      }
      return CheckFilledIcon;
    }
    return CheckEmptyIcon;
  })();

  const Element = showLabel ? 'label' : 'div';

  return (
    <Element className="display--inlineBlock">
      <CustomCheckboxContainer
        checked={rest.checked != null ? rest.checked : checked}
        onChange={onChange}
        className="fdsCheckboxContainer"
      >
        <CustomCheckboxInput {...rest} aria-label={!showLabel ? label : undefined} />
        <span aria-hidden style={{ pointerEvents: 'none' }}>
          <CheckboxIcon size="xs" color={checked ? FDS.COLOR_BLUE : undefined} />
        </span>
      </CustomCheckboxContainer>
      {showLabel && <span className="padding--left--half">{label}</span>}
    </Element>
  );
};

Checkbox.propTypes = {
  /**
   * Setting any one of the 3 values here will turn component into controlled component.
   * */
  checked: PropTypes.oneOf([true, false, 'mixed']),

  /** Label used for `label` element */
  label: PropTypes.string.isRequired,

  /** `true` checks the checkbox by default */
  defaultChecked: PropTypes.bool,

  /** Disables form field when `true` */
  disabled: PropTypes.bool,

  onChange: PropTypes.func,

  /** Controls whether the label surfaces as a text label, or is hidden and applied as an aria-label */
  showLabel: PropTypes.bool,
};

export default Checkbox;
