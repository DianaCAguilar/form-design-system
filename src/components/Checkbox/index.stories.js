import React from 'react';
import Checkbox from '.';

const Template = (args) => <Checkbox {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  name: 'spam',
  label: 'I agree to receive spam',
};

export const managingCheckedState = Template.bind({});

managingCheckedState.args = {
  defaultChecked: true,
  name: 'default-checked',
  label: 'Checked by default',
};

managingCheckedState.parameters = {
  docs: {
    description: {
      story:
        '`Checkbox` will render as unchecked by default. Use the `defaultChecked` boolean prop to set the initial state to checked.',
    },
  },
};

export const preventingUserInteraction = () => (
  <>
    <div>
      <Checkbox disabled name="disabled-unchecked" label="Disabled unchecked" />
    </div>
    <div>
      <Checkbox
        disabled
        defaultChecked
        name="disabled-checked"
        label="Disabled checked"
      />
    </div>
  </>
);

preventingUserInteraction.parameters = {
  docs: {
    description: {
      story:
        'Use the `disabled` prop to prevent the user from changing the checked state of the checkbox.',
    },
  },
};

export const checkboxWithoutLabel = Template.bind({});
checkboxWithoutLabel.args = {
  name: 'no-label',
  label: "I don't have a label",
  showLabel: false,
};

checkboxWithoutLabel.parameters = {
  docs: {
    description: {
      story:
        'You can prevent the label from showing by setting `showLabel` to `false`. The value of the `label` prop will be used to set an ARIA attribute.',
    },
  },
};

export const indeterminateCheckboxes = Template.bind({});
indeterminateCheckboxes.args = {
  indeterminate: true,
  name: 'notsureif',
  label: 'Indeterminate checkbox',
};

indeterminateCheckboxes.parameters = {
  docs: {
    description: {
      story:
        'Passing the `indeterminate` boolean prop will set `indeterminate` attribute of the underlying input element and render the `CheckIndeterminate` icon when the checkbox is checked.',
    },
  },
};

export const readingTheValueOfACheckbox = Template.bind({});
readingTheValueOfACheckbox.args = {
  label: 'Option one',
};

readingTheValueOfACheckbox.parameters = {
  docs: {
    description: {
      story:
        ' The `onChange` prop will be spread on the `input` element the `Checkbox` component renders. Read the `checked` attribute of the event `currentTarget`.',
    },
  },
};

export default {
  component: Checkbox,
  title: 'components/Checkbox',
  argTypes: { onChange: { action: 'onChange' } },
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
      },
      description: {
        component: 'Uncontrolled custom checkbox component with optional label.',
      },
    },
  },
};
