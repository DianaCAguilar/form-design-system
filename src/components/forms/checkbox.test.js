import React from 'react';
import uuidv4 from 'uuid/v4';
import { shallow } from 'enzyme';

// uuid will break snapshots, so we must mock it.
jest.mock('uuid/v4', () => jest.fn().mockReturnValue('mock-universal-unique-identifier'));

import Checkbox from './Checkbox';

describe('Checkbox component', () => {

  it('matches snapshot (without label)', () => {
    const wrapper = shallow(<Checkbox name="snapshot_checkbox" label="snapshot checkbox" onChange={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (with label)', () => {
    const wrapper = shallow(<Checkbox label="Snapshot Label" name="snapshot_checkbox" onChange={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (disabled)', () => {
    const wrapper = shallow(<Checkbox disabled label="Snapshot Label" name="snapshot_checkbox" onChange={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('fires change callback when checking', () => {
    const changeFn = jest.fn();
    const input = shallow(<Checkbox name="check-callback-test" onChange={changeFn} />)
      .find('input');

    expect(changeFn).not.toHaveBeenCalled();
    input.simulate('change', { target: { checked: true } })
    expect(changeFn).toHaveBeenCalled();
  });

  it('fires change callback when unchecking', () => {
    const changeFn = jest.fn();
    const input = shallow(<Checkbox checked name="uncheck-callback-test" onChange={changeFn} />)
      .find('input');

    expect(changeFn).not.toHaveBeenCalled();
    input.simulate('change', { target: { checked: false } })
    expect(changeFn).toHaveBeenCalled();
  });

  describe('`checked` prop', () => {

    it('sets initial state to unchecked when `checked` prop is NOT passed', () => {
      const input = shallow(<Checkbox name="unchecked-initial" />)
        .find('input');
      expect(input.prop('checked')).toBe(false);
    });

    it('sets initial state to checked when `checked` prop is passed', () => {
      const input = shallow(<Checkbox name="checked-initial" checked />)
        .find('input');
      expect(input.prop('checked')).toBe(true);
    });

  });

  describe('checkbox icons', () => {

    it('uses unchecked icon when unchecked', () => {
      const wrapper = shallow(<Checkbox name="unchecked-icon" />);
      expect(wrapper.exists('CheckEmptyIcon')).toBe(true);
    });

    it('uses checked icon when checked', () => {
      const wrapper = shallow(<Checkbox name="checked-icon" checked />);
      expect(wrapper.exists('CheckFilledIcon')).toBe(true);
    });

    it('uses indeterminate check icon when `indeterminate` prop is passed and is checked', () => {
      const wrapper = shallow(<Checkbox name="indeterminate-initial" checked indeterminate />);
      expect(wrapper.exists('CheckIndeterminateIcon')).toBe(true);
    });

  });

});
