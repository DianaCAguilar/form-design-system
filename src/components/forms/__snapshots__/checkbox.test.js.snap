// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component matches snapshot (disabled) 1`] = `
<div
  className="fdsCheckbox fdsCheckbox--disabled"
>
  <input
    checked={false}
    className="media--xs"
    disabled={true}
    id="snapshot_checkbox"
    name="snapshot_checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="flush--bottom"
    htmlFor="snapshot_checkbox"
  >
    <div
      aria-checked={false}
      role="checkbox"
    >
      <CheckEmptyIcon
        size="xs"
      />
    </div>
    <span
      className="padding--left--half"
    >
      Snapshot Label
    </span>
  </label>
</div>
`;

exports[`Checkbox component matches snapshot (with label) 1`] = `
<div
  className="fdsCheckbox"
>
  <input
    checked={false}
    className="media--xs"
    disabled={false}
    id="snapshot_checkbox"
    name="snapshot_checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="flush--bottom"
    htmlFor="snapshot_checkbox"
  >
    <div
      aria-checked={false}
      role="checkbox"
    >
      <CheckEmptyIcon
        size="xs"
      />
    </div>
    <span
      className="padding--left--half"
    >
      Snapshot Label
    </span>
  </label>
</div>
`;

exports[`Checkbox component matches snapshot (without label) 1`] = `
<div
  className="fdsCheckbox"
>
  <input
    checked={false}
    className="media--xs"
    disabled={false}
    id="snapshot_checkbox"
    name="snapshot_checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="flush--bottom"
    htmlFor="snapshot_checkbox"
  >
    <div
      aria-checked={false}
      role="checkbox"
    >
      <CheckEmptyIcon
        size="xs"
      />
    </div>
  </label>
</div>
`;
